src/sdk.rs

```diff
@@ -61,33 +60,50 @@ impl Sdk {
         let builder = tonic::transport::channel::Channel::builder(addr)
             .keep_alive_timeout(keep_alive.unwrap_or_else(|| Duration::from_secs(30)));
 
-        let channel = builder.connect().await?;
-        let mut client = SdkClient::new(channel.clone());
-
-        #[cfg(feature = "player-tracking")]
-        let alpha = AlphaClient::new(channel);
-
-        tokio::time::timeout(
-            handshake_timeout.unwrap_or_else(|| Duration::from_secs(30)),
+        let (client, game_server, _channel) = tokio::time::timeout(
+            connect_timeout.unwrap_or_else(|| Duration::from_secs(30)),
             async {
-                let mut connect_interval = tokio::time::interval(Duration::from_millis(100));
+                let mut connect_interval = tokio::time::interval(Duration::from_millis(1));
 
-                loop {
+                let channel = loop {
                     connect_interval.tick().await;
 
-                    if client.get_game_server(empty()).await.is_ok() {
-                        break;
+                    // It would be nice to differentiate between transient errors
+                    // (eg, agones sidecar is still initializing) and hard errors
+                    // but tonic's error doesn't really allow good introspection
+                    // so we just retry until we succeed or timeout
+                    if let Ok(channel) = builder.connect().await {
```

Comment:
Tonic has https://docs.rs/tonic/0.5.2/tonic/transport/channel/struct.Endpoint.html#method.connect_timeout but I think this is the best approach if you wanna combine connecting and `client.get_game_server` into one future with a timeout.